swfiles/sw_fitpowder.m

```diff
@@ -456,18 +504,10 @@ function estimate_scale_factor(obj)
         end
 
         function estimate_constant_background(obj)
-            ysort = sort(obj.y(obj.y < mean(obj.y, 'omitnan')), 'descend');
-            bg = ysort(int32(numel(ysort)/2)); % default to median
-            prev_skew = inf;
-            for ipt = 1:numel(ysort)
-                this_mean = mean(ysort(ipt:end));
-                this_skew = mean((ysort(ipt:end) - this_mean).^3)/(std(ysort(ipt:end)).^3);
-                if this_skew < 0 || this_skew > prev_skew
-                    bg = this_mean;
-                    break
-                else
-                    prev_skew = this_skew;
-                end
+            if isempty(obj.ibg)
+                bg = obj.find_indices_and_mean_of_bg_bins();
+            else
+                bg = mean(obj.y(obj.ibg));
             end
```

Comment:
I don't know how likely this is, but in `find_indices_and_mean_of_bg_bins` if the skew keeps decreasing but never gets below `0` then `ibg` is just the lowest value point, _but_ the value of `bg` is the default which is the median of all values - however, subsequent calls to `estimate_constant_background` will change `bg` to the mean of `ibg` (e.g. set it to the lowest finite value in the dataset...

Is this edge case likely to happen? (should we guard against it in `find_indices_and_mean_of_bg_bins` somehow?)

Reply:
Good spot, sorry I missed that. I don't think it's likely but we should guard against it. I have a couple of suggestions:

1. One option is not to have a default background - i.e. in your scenario above the background will then just be the bin with the minimum intensity (which could be zero), then at least the `ibg` and the background are consistent. In this case though we should also print a warning I think? 
2. The other option is to say the background estimation failed and don't set `ibg` - maybe have another flag to denote failed background estimation that we can check in the other methods that use `ibg`
3. I could also set `ibg` to be consistent with the default background (median of all points less than the mean of `y`) - i.e. `obj.ibg = iseed(isort(int32(numel(isort)/2):end))

I'm leaning towards (2) and when trying to fit the background check how many points there are so as not to attempt fitting if too few points.

Reply:
I think option (2) is also best, with option (1) a close second. If you set `bg` in line 702 to be `NaN` instead of the median, you can check at the end after line 714 if `bg` is still `NaN` after the loop and then emit an error there and only set `obj.ibg` if `bg` is not NaN - thus if the background finding algorithm fails `obj.ibg` would stay empty `[]` and this routine would be called again each time `fit_background` / `estimate_constant_background` is called... (or as you say we could have a hidden property as a flag).